File: epi_judge_java/epi/TreeInorder.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class TreeInorder {

    @EpiTest(testDataFile = "tree_inorder.tsv")
    public static List<Integer> inorderTraversal(BinaryTreeNode<Integer> tree) {
        List<Integer> result = new ArrayList<>();

        Deque<BinaryTreeNode<Integer>> stack = new ArrayDeque<>();
        var runner = tree;

        while (!stack.isEmpty() || runner != null) {
            if (runner != null) {
                stack.push(runner);
                runner = runner.left;
            } else {
                runner = stack.pop();
                result.add(runner.data);
                runner = runner.right;
            }
        }

        return result;
    }


    private static class NodeAndState {
        public BinaryTreeNode<Integer> node;
        public Boolean leftSubtreeTraversed;

        public NodeAndState(BinaryTreeNode<Integer> node,
                            Boolean leftSubtreeTraversed) {
            this.node = node;
            this.leftSubtreeTraversed = leftSubtreeTraversed;
        }
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "TreeInorder.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/tree_inorder.cc
#include <memory>
#include <vector>

#include "binary_tree_node.h"
#include "test_framework/generic_test.h"

using std::unique_ptr;
using std::vector;
using std::stack;

vector<int> InorderTraversal(const unique_ptr<BinaryTreeNode<int>> &tree) {
    vector<int> result;

    stack<const BinaryTreeNode<int> *> st;
    const auto *curr = tree.get();

    while (!st.empty() || curr) {
        if (curr) {
            st.push(curr);
            curr = curr->left.get();
        } else {
            curr = st.top();
            st.pop();
            result.push_back(curr->data);
            curr = curr->right.get();
        }
    }

    return result;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"tree"};
    return GenericTestMain(args, "tree_inorder.cc", "tree_inorder.tsv",
                           &InorderTraversal, DefaultComparator{}, param_names);
}
